### ISSUE TITLE
 bird: I/O loop cycle took

### ISSUE THREAD

**Issue by ShawnLeung87** [AltraMayor/gatekeeper#660]
An abnormal io log appears in bird, and a reload appears in gatekeeper. After reload, many flows were rejected and could not pass.
Sep 29 16:22:20 77-GK-2 bird: I/O loop cycle took 173558 ms for 7 events
Sep 29 16:22:48 77-GK-2 bird: I/O loop cycle took 27523 ms for 5 events
Sep 29 16:23:14 77-GK-2 bird: I/O loop cycle took 20373 ms for 1 events
Sep 29 16:28:21 77-GK-2 bird: I/O loop cycle took 105873 ms for 5 events
Sep 29 16:33:49 77-GK-2 bird: I/O loop cycle took 170328 ms for 7 events
Sep 29 16:37:41 77-GK-2 bird: I/O loop cycle took 48777 ms for 4 events
Sep 29 16:40:43 77-GK-2 bird: I/O loop cycle took 13872 ms for 2 events
Sep 29 16:45:10 77-GK-2 bird: I/O loop cycle took 119808 ms for 6 events
Sep 29 16:50:41 77-GK-2 bird: I/O loop cycle took 188279 ms for 7 events
Sep 29 16:54:44 77-GK-2 bird: I/O loop cycle took 101038 ms for 6 events
Sep 29 16:59:48 77-GK-2 bird: I/O loop cycle took 62101 ms for 5 events
Sep 29 17:01:02 77-GK-2 bird: I/O loop cycle took 74442 ms for 6 events
Sep 29 17:06:04 77-GK-2 bird: I/O loop cycle took 98967 ms for 6 events
Sep 29 17:11:12 77-GK-2 bird: I/O loop cycle took 173759 ms for 7 events
Sep 29 17:13:35 77-GK-2 bird: I/O loop cycle took 12505 ms for 4 events
Sep 29 17:16:24 77-GK-2 bird: I/O loop cycle took 80281 ms for 5 events
Sep 29 17:21:54 77-GK-2 bird: I/O loop cycle took 124891 ms for 6 events
Sep 29 17:22:27 77-GK-2 bird: I/O loop cycle took 23494 ms for 2 events
Sep 29 17:26:00 77-GK-2 bird: I/O loop cycle took 69845 ms for 5 events
Sep 29 17:31:32 77-GK-2 bird: I/O loop cycle took 34428 ms for 4 events
Sep 29 17:31:57 77-GK-2 bird: I/O loop cycle took 23200 ms for 2 events
Sep 29 17:36:52 77-GK-2 bird: I/O loop cycle took 62082 ms for 5 events
Sep 29 17:37:34 77-GK-2 bird: I/O loop cycle took 30275 ms for 3 events
Sep 29 17:38:09 77-GK-2 bird: I/O loop cycle took 14942 ms for 4 events
Sep 29 17:48:14 77-GK-2 bird: I/O loop cycle took 20316 ms for 4 events

**Comment by AltraMayor** [comment]
Hi @ShawnLeung87,

1. How often is this problem happening?

2. Can you reproduce it in your test environment?

3. Could you run the script `gkctl/scripts/summarize_fibs.lua` and post the output here? If you can reproduce this problem, what's the script's output **while** the problem is going on?

4. What did you find in the Gatekeeper's log during the issue?

5. Do you have scripts that repeatedly connect to the dynamic configuration of Gatekeeper? If so, what do they do?

6. How many routers do you have in front of the Gatekeeper server?

**Comment by ShawnLeung87** [comment]
I have not enabled the use of ipv6 and have not configured ipv6. However, Gatekeeper has this log.（When an exception log above bird appears）
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 15, f.v6.src = 0x670dde7c000000000000000000080000, f.v6.dst = 0x2f6d551d67643e0
30000000002000000} at index 98909343: gk_del_flow_entry_with_key(): flow_key is invalid at position 98909343. rte_hash_free_key_w
ith_position() returned 0 (i.e. Success). Logging and removing flow entry...
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x0000000000000000
0000000000000000} at index 98909343: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at
: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x0000000000000000
0000000000000000} at index 98909343: gk_del_flow_entry_at_pos(): flow key is invalid; logging and removing flow entry...
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x0000000000000000
0000000000000000} at index 98909343: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at
: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 39840, f.v6.src = 0x2bf093e1000000000000000000083880, f.v6.dst = 0xdaf759d19ace
efe6e255000045000000} at index 106963116: gk_del_flow_entry_with_key(): flow_key is invalid at position 106963116. rte_hash_free_
key_with_position() returned 0 (i.e. Success). Logging and removing flow entry...
[2023-10-11 13:23:59] GK/7: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x0000000000000000
0000000000000000} at index 106963116: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen

**Comment by AltraMayor** [comment]
This problem is happening on only one of your three production Gatekeeper servers, right?

The problem seems to be a corruption of the flow table at instance GK/7. I see two possibilities here: (1) a memory problem, or (2) a bug or problem somewhere. Before we delve into investigating the problem further, could you move the Gatekeeper server presenting this problem offline and test its memory with [MemTest86](https://www.memtest86.com/)? Ubuntu should have a package for MemTest86. Also, consider adding a Gatekeeper server in place of the one you are removing to keep your deployment as it is.

**Comment by ShawnLeung87** [comment]
It's not that there's a problem with one unit, it's happening to all three units. This problem will recur when I direct small packet attack traffic to one of the machines exceeding 3Gbps.

**Comment by ShawnLeung87** [comment]
Okay, let's test whether there is a problem with the memory first. We will use the MemTest86 tool to test memory offline

**Comment by ShawnLeung87** [comment]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 113062761: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 113062761: gk_del_flow_entry_at_pos(): flow key is invalid; logging and removing flow entry...
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 113062761: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 23456, f.v6.src = 0x2bf1fc24000000000000000000080000, f.v6.dst = 0x2ddf2f4d6790c577647f0000a05c7fbd} at index 106239377: gk_del_flow_entry_with_key(): flow_key is invalid at position 106239377. rte_hash_free_key_with_position() returned 0 (i.e. Success). Logging and removing flow entry...
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 106239377: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 106239377: gk_del_flow_entry_at_pos(): flow key is invalid; logging and removing flow entry...
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 106239377: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 43208, f.v6.src = 0x67d73085000000000000000000086ec7, f.v6.dst = 0x2f6b699267643c571056000045000000} at index 94156245: gk_del_flow_entry_with_key(): flow_key is invalid at position 94156245. rte_hash_free_key_with_position() returned 0 (i.e. Success). Logging and removing flow entry...
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 94156245: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 94156245: gk_del_flow_entry_at_pos(): flow key is invalid; logging and removing flow entry...
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 0, f.v6.src = 0x00000000000000000000000000000000, f.v6.dst = 0x00000000000000000000000000000000} at index 94156245: NOT in use [state: GK_REQUEST (0), flow_hash_value: 0x0, expire_at: 0x0, last_packet_seen_at: 0x0, last_priority: 0, allowance: 0, grantor_ip: NULL FIB entry]
[2023-10-13 00:23:58] GK/10: INVALID Flow {proto = 30, f.v6.src = 0x67c4cc03000000000000000000080000, f.v6.dst = 0xce6e

**Comment by ShawnLeung87** [comment]
i40e_flow_set_fdir_inset(): Invalid input set 
I encountered this error when restarting gatekeeper. Is it serious?

**Comment by ShawnLeung87** [comment]
i40e_check_write_global_reg(): i40e device 0000:5e:00.1 changed global register [0x0026774c]. original: 0x0007fffe, new: 0x0007fff8
i40e_check_write_global_reg(): i40e device 0000:5e:00.1 changed global register [0x0026775c]. original: 0x0007fffe, new: 0x0007fff8
[2023-10-12 11:16:51] Main/0: check_port_rss(front): port 1 (0000:5e:00.1) only supports RSS hash functions 0x7ef8, but Gatekeeper asks for 0xafbc
[2023-10-12 11:16:51] Main/0: check_port_rss(back): port 0 (0000:5e:00.0) only supports RSS hash functions 0x7ef8, but Gatekeeper asks for 0xafbc
[2023-10-12 11:16:51] Main/0: lls: calculate_mempool_config_para: total_pkt_burst = 2144 packets, total_rx_desc = 640 descriptors, total_tx_desc = 256 descriptors, max_num_pkt = 3040 packets, num_mbuf = 4095 packets.
[2023-10-12 11:16:51] Main/0: gk: calculate_mempool_config_para: total_pkt_burst = 612 packets, total_rx_desc = 640 descriptors, total_tx_desc = 256 descriptors, max_num_pkt = 1508 packets, num_mbuf = 2047 packets.
[2023-10-12 11:18:14] Main/0: ggu: calculate_mempool_config_para: total_pkt_burst = 32 packets, total_rx_desc = 640 descriptors, total_tx_desc = 256 descriptors, max_num_pkt = 928 packets, num_mbuf = 1023 packets.
[2023-10-12 11:18:14] Main/0: cps: calculate_mempool_config_para: total_pkt_burst = 4256 packets, total_rx_desc = 640 descriptors, total_tx_desc = 256 descriptors, max_num_pkt = 5152 packets, num_mbuf = 8191 packets.
[2023-10-12 11:18:18] Main/0: ethertype_flow_add(front): EtherType=0x806 flow supported
[2023-10-12 11:18:18] Main/0: ethertype_flow_add(back): EtherType=0x806 flow supported
[2023-10-12 11:18:18] Main/0: Back interface link speed: 5000000000 bytes per second
[2023-10-12 11:18:18] Main/0: Cycles per byte (19.155015) represented as a rational: 148777 / 7767
i40e_flow_set_fdir_inset(): Invalid input set
[2023-10-12 11:18:18] Main/0: ipv4_flow_add(back, DstIP=10.11.10.196 UDP SrcPort=41120/0xffff DstPort=45232/0xffff): cannot validate IPv4 flow, errno=22 (Invalid argument), rte_flow_error_type=13: Invalid ETH item
[2023-10-12 11:18:18] Main/0: Cannot register IPv4 flow on the back interface; falling back to software filters
i40e_flow_set_fdir_inset(): Invalid input set
[2023-10-12 11:18:18] Main/0: ipv4_flow_add(front, DstIP=10.12.11.3 TCP SrcPort=0/0x0 DstPort=0/0x0): cannot validate IPv4 flow, errno=22 (Invalid argument), rte_flow_error_type=13: Invalid ETH item
[2023-10-12 11:18:18] Main/0: Cannot register IPv4 flow on the front interface; falling back to software filters
i40e_flow_set_fdir_inset(): Invalid input set

**Comment by AltraMayor** [comment]
> i40e_flow_set_fdir_inset(): Invalid input set I encountered this error when restarting gatekeeper. Is it serious?

The original issue is already complicated, so please open another issue to discuss this second problem. I can't address both issues simultaneously; mixing them here would make this investigation harder.

I'm waiting on the memory test to decide how to move forward. I understand that you've found the problem in other production Gatekeepers but have yet to be able to reproduce the problem in your test environment. So, we need to gradually scope the problem to save time on something we can check at the beginning.

**Comment by ShawnLeung87** [comment]
<img width="592" alt="WechatIMG4898" src="https://github.com/AltraMayor/gatekeeper/assets/104405912/c147aacc-c279-46b1-89e6-10e39dc07754">

We conducted 4 rounds of memory checks and found no hardware abnormalities. The MEMtestX86 prompt was pass.
I looked at the gatekeeper's exception log. In the case of a small packet (64-byte packet) reflection attack, when the source IP exceeds 500 /24 prefix, the bird of gatekeeper declares that our company's IP (destination IP prefix /24) exceeds 5 /24 The prefix and flow are most prone to damage. We check the flow index id, and the maximum reaches 100 million. I'm wondering if I can filter packets based on the source IP before creating the flow table, and reject packets before creating the flow? We have packet-based analysis software that can analyze most malicious reflection IP segments. If combined with packet filtering, the flow table performance pressure on gatekeeper can be greatly reduced.

**Comment by AltraMayor** [comment]
Thank you for having taken the time to test the memory. Having eliminated this possibility early on will help us with the solution to this problem.

I've written the patch [no_full_scan.TXT](https://github.com/AltraMayor/gatekeeper/files/13073059/no_full_scan.TXT) that you can apply against version `v1.1.0`. This patch avoids a full table scan of the flow table of an instance of the GK block. This is a temporary solution, but it'll prevent disrupting Bird. Gatekeeper will still log problems in the flow table. Before deploying it in production, please test your Gatekeeper with this patch in your test environment. And deploy it in production one machine at a time, waiting a couple of days between machines.

If the patch is not sufficient to contain the problem until we get to the final solution, we can explore other temporary solutions, such as your suggestions.

The final solution will be the new data structure we've been writing for the flow table of version 1.2 of Gatekeeper. We should have that code available for you to test in about a month or earlier.

**Comment by ShawnLeung87** [comment]
We simulated a small packet (64-byte) attack and a multi-source IP environment in the test scenario. The flow table is also damaged. We also simulated the gatekeeper in the default gateway mode, and the same situation occurred.

**Comment by ShawnLeung87** [comment]
> Thank you for having taken the time to test the memory. Having eliminated this possibility early on will help us with the solution to this problem.
> 
> I've written the patch [no_full_scan.TXT](https://github.com/AltraMayor/gatekeeper/files/13073059/no_full_scan.TXT) that you can apply against version `v1.1.0`. This patch avoids a full table scan of the flow table of an instance of the GK block. This is a temporary solution, but it'll prevent disrupting Bird. Gatekeeper will still log problems in the flow table. Before deploying it in production, please test your Gatekeeper with this patch in your test environment. And deploy it in production one machine at a time, waiting a couple of days between machines.
> 
> If the patch is not sufficient to contain the problem until we get to the final solution, we can explore other temporary solutions, such as your suggestions.
> 
> The final solution will be the new data structure we've been writing for the flow table of version 1.2 of Gatekeeper. We should have that code available for you to test in about a month or earlier.

I will follow this patch to test

**Comment by ShawnLeung87** [comment]
After the patch is updated, the full table scan is no longer performed, and bird no longer encounters exceptions. However, when a small packet attack occurs and the number of flow tables in each instance accumulates to more than 200 million, the recycling cannot keep up with the new creation, resulting in inaccessible IP addresses.

**Comment by AltraMayor** [comment]
Thank you for this information. Could you add more details? Did it happen in your test environment or production? If it happened in your test environment, could you describe the test you performed? What's the value of the parameter `flow_ht_size` in `lua/gk.lua`?

We are still working on the final solution.

**Comment by ShawnLeung87** [comment]
gk.lua parameter .Enable 4 cores.Policy expiration time is set to 720 seconds.
local mailbox_max_entries_exp = 14
        local mailbox_mem_cache_size = 0
        local mailbox_burst_size = 32
        local log_ratelimit_interval_ms = 5000
        local log_ratelimit_burst = 100
        local max_pkt_burst_front = 32
        local max_pkt_burst_back = 32

        local flow_ht_size = 250000000
        local flow_table_scan_iter = 0
        local scan_del_thresh = flow_ht_size * 0.1

        local max_num_ipv4_rules = 1600000
        local num_ipv4_tbl8s = 200000
        local max_num_ipv6_rules = 400000
        local num_ipv6_tbl8s = 100000
        local max_num_ipv6_neighbors = 65536

        local basic_measurement_logging_ms = 60 * 1000 -- (1 minute)

        local front_icmp_msgs_per_sec = 1000
        local front_icmp_msgs_burst = 50
        local back_icmp_msgs_per_sec = 1000
        local back_icmp_msgs_burst = 50

        local fib_dump_batch_size = 32

**Comment by ShawnLeung87** [comment]
Randomly, some IPs have no flow records. As a result, the IP cannot be accessed.
When the attack occurred for 3 hours and each core flow table used 200 million entries, the phenomenon became serious. A random check of a pair of src ip and dst ip that was inaccessible showed that the gatekeeper did not have any related flow records.

**Comment by ShawnLeung87** [comment]
If the problem of time-consuming exceptions caused by hash table storage and search can be solved, gatekeeper will be more effective in defending against attacks by TCP small packets (64 bytes). We have compared most of the protection software on the market, and they are not as effective as gatekeeper. They are all based on stateless packet rate limiting. All flows share a rate-limiting bucket, which will cause the destination server side to be unable to access the Internet (due to congestion). Based on statefulness, it can be well analyzed whether the flow is abnormal, thereby rejecting the flow without affecting the normal flow.

**Comment by AltraMayor** [comment]
The coming new flow table will not have the memory exceptions, but its development is slow.

There are two things that you can do to deal with this problem In the meaning while:
1. Reducing the expiration of your policy decision (i.e., parameter `cap_expire_sec`) from 720s (= 12 minutes) to 300s (= 5 minutes). This should be fine because policy decisions in use are renewed before expiring. However, a shorter expiration will enable your Gatekeeper servers to reclaim flow entries not in use quicker.
2. Adding more Gatekeeper servers to your deployment. The symptoms you've described and the fact that each instance of your flow tables has more than 80% (= 200 million entries / (flow_ht_size = 250000000)) of its capacity already allocated point out that you don't have enough flow table capacity for your deployment. This problem is analogous to having a single 40Gbps Gatekeeper server handle a front link of 100Gbps. Given that you often face these attacks, I recommend doubling the number of Gatekeeper servers you currently have in production.

**Comment by AltraMayor** [comment]
Hi @ShawnLeung87,

We can now start the tests of the new flow table. You need to set up your test environment to run with branch `v1.2.0-dev` available on the repository. The first goal is to have a **repeatable** way to reproduce the issue with that version of the code. That version of Gatekeeper uses the save flow table of `v1.0.0`, so you should be able to reach this goal easily. Once you get this goal, could you post **all** the logs that the Gatekeeper server produced during the test? This log will be our baseline.

After posting the baseline, we'll move to the second goal: repeating the same test with the new flow table.
